optuna/progress_bar.py

```diff
@@ -59,22 +76,35 @@ def _init_valid(self) -> None:
         optuna_logging.disable_default_handler()
         optuna_logging._get_library_root_logger().addHandler(_tqdm_handler)
 
-    def update(self, elapsed_seconds: Optional[float]) -> None:
+    def update(self, elapsed_seconds: float) -> None:
         """Update the progress bars if ``is_valid`` is :obj:`True`.
 
         Args:
             elapsed_seconds:
                 The time past since :func:`~optuna.study.Study.optimize` started.
         """
+
         if self._is_valid:
-            self._progress_bar.update(1)
-            if self._timeout is not None and elapsed_seconds is not None:
-                self._progress_bar.set_postfix_str(
-                    "{:.02f}/{} seconds".format(elapsed_seconds, self._timeout)
-                )
+            if self._n_trials is not None:
+                self._progress_bar.update(1)
+                if self._timeout is not None:
+                    self._progress_bar.set_postfix_str(
+                        "{:.02f}/{} seconds".format(elapsed_seconds, self._timeout)
+                    )
+
+            else:
+                assert self._timeout is not None
+                time_diff = elapsed_seconds - self._last_elapsed_seconds
+                if elapsed_seconds > self._timeout:
+                    # Clip elapsed time to avoid tqdm warnings.
+                    time_diff -= elapsed_seconds - self._timeout
+
+                self._progress_bar.update(time_diff)
+                self._last_elapsed_seconds = elapsed_seconds
```

Comment:
This might be out-of-scope of this PR, but I'm concerned that we need to rewrite this logic when we support multi-threading. As described in #2957, Optuna plans to enable the progress bar with `n_jobs!=1` and #3138 conducts on the change.

I found that this code did not work as expected.

1. The percentage can exceed 100%.
```console
[I 2021-12-01 18:16:01,496] Trial 78 finished with value: 0.00023993382411588578 and parameters: {'x': 0.00023993382411588578}. Best is trial 51 with value: 5.7434369678507294e-05.                                                                                    
 98%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▍     | 00:03/00:04time_diff=0.13739999999999997, elapsed_seconds=3.886893, last_elapsed_seconds=3.749493
/home/yanase/pfn/code/optuna-review/.venv/lib/python3.8/site-packages/tqdm/std.py:533: TqdmWarning: clamping frac to range [0, 1]
  full_bar = Bar(frac,
[I 2021-12-01 18:16:01,513] Trial 93 finished with value: 0.0693096165017975 and parameters: {'x': 0.0693096165017975}. Best is trial 51 with value: 5.7434369678507294e-05.                                                                                            
/home/yanase/pfn/code/optuna-review/.venv/lib/python3.8/site-packages/tqdm/std.py:533: TqdmWarning: clamping frac to range [0, 1]
  full_bar = Bar(frac,
101%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 00:03/00:04
```

2. The percentage can decrease after `elapsed_seconds` exceeds `timeout`.

```console
101%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 00:04/00:04time_diff=-0.22978699999999996, elapsed_seconds=4.236397, last_elapsed_seconds=4.229787
[I 2021-12-01 18:16:01,893] Trial 86 finished with value: 0.04476754994720062 and parameters: {'x': 0.04476754994720062}. Best is trial 51 with value: 5.7434369678507294e-05.                                                                                          
 95%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋            | 00:04/00:04time_diff=-0.2363970000000002, elapsed_seconds=4.28378, last_elapsed_seconds=4.236397
[I 2021-12-01 18:16:01,909] Trial 89 finished with value: 0.0032808017066235364 and parameters: {'x': 0.0032808017066235364}. Best is trial 51 with value: 5.7434369678507294e-05.                                                                                      
 89%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏                          | 00:04/00:04time_diff=-0.28378000000000014, elapsed_seconds=4.300264, last_elapsed_seconds=4.28378
[I 2021-12-01 18:16:01,923] Trial 95 finished with value: 0.000660676902939195 and parameters: {'x': 0.000660676902939195}. Best is trial 51 with value: 5.7434369678507294e-05.                                                                                        
 82%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋                                            | 00:04/00:04
```


Do you think we can extend this logic to support multi-threading?

Reply:
I think the problem here is that each thread keeps its own timing, but references the same global progress bar. Without common source of time elapsed, it produces those artifacts above.

By looking at current implementation of `_optimize`, I think we can fix this by moving progress bar updates to executor context. In pseudocode:

```python
[...]
progress_bar = pbar_module._ProgressBar(show_progress_bar and n_jobs == 1, n_trials, timeout)
[...]

with ThreadPoolExecutor(max_workers=n_jobs) as executor:
    for n_submitted_trials in itertools.count():
        if study._stop_flag:
            break

        if (
            timeout is not None
            and (datetime.datetime.now() - time_start).total_seconds() > timeout
        ):
            break

        if n_trials is not None and n_submitted_trials >= n_trials:
            break

        if len(futures) >= n_jobs:
            completed, futures = wait(futures, return_when=FIRST_COMPLETED)
            for f in completed:
                # pbar update is handled in executor context, common for all tasks
                progress_bar.update((datetime.datetime.now() - time_start).total_seconds())
                f.result()

        futures.add(
            executor.submit(
                _optimize_sequential,
                study,
                func,
                1,
                timeout,
                catch,
                callbacks,
                gc_after_trial,
                True,
                time_start,
                None,  # we do not pass progress bar reference to task
            )
        )
```

Reply:
So I wrote a [quick and dirty](https://github.com/xadrianzetx/optuna/blob/3582b131f1d493d6cdf51265dfe2141c1385790d/optuna/study/_optimize.py#L84-L123) implementation of the above, which _appears_ to work. [Here's a collab notebook](https://colab.research.google.com/drive/1r88auxJBRqokHHaOh_RuWgZ6kcLsL_4s?usp=sharing) so you can try it out. The change could be introduced in #3138 and this PR is compatible with it, apart from small `elapsed` formatting issue.